src/agent/src/util.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::sync::watch::Receiver;
+
+// Size of I/O read buffer
+const BUF_SIZE: usize = 8192;
+
+// Interruptable I/O copy using readers and writers
+// (an interruptable version of "io::copy()").
+pub async fn interruptable_io_copier<R: Sized, W: Sized>(
+    mut reader: R,
+    mut writer: W,
+    mut shutdown: Receiver<bool>,
+) -> io::Result<u64>
+where
+    R: tokio::io::AsyncRead + Unpin,
+    W: tokio::io::AsyncWrite + Unpin,
+{
+    let mut total_bytes: u64 = 0;
+
+    let mut buf: [u8; BUF_SIZE] = [0; BUF_SIZE];
+
+    loop {
+        tokio::select! {
+            _ = shutdown.changed() => {
+                eprintln!("INFO: interruptable_io_copier: got shutdown request");
+                break;
```

Comment:
Just a question: is there only one way to exit this function when receiving `shutdown`? Other conditions like reader closed should `continue` or `break` at line 38 or 44?  Maybe the answer depends on the calling side.

Reply:
Good catch @liubin! I've updated the I/O copier to have the same semantics as `std::io::copy()` for a read count of zero. I've also updated the tests to handle more scenarios. Sigh... It would be useful if tokio provided this for us... :smile: